test/Microsoft.Azure.Devices.Edge.Test.Common/EdgeModule.cs

```diff
@@ -130,31 +130,71 @@ public Task UpdateDesiredPropertiesAsync(object patch, CancellationToken token)
         {
             return Profiler.Run(
-                () =>
-                {
-                    return Retry.Do(
-                        async () =>
-                        {
-                            Twin twin = await this.iotHub.GetTwinAsync(this.deviceId, this.Id, token);
-                            return twin.Properties.Reported;
-                        },
-                        reported =>
-                        {
-                            JObject expected = JObject.FromObject(expectedPatch)
-                                .Value<JObject>("properties")
-                                .Value<JObject>("reported");
-                            return expected.Value<JObject>().All<KeyValuePair<string, JToken>>(
-                                prop => reported.Contains(prop.Key) && reported[prop.Key] == prop.Value);
-                        },
-                        null,
-                        TimeSpan.FromSeconds(5),
-                        token);
-                },
+                () => this.WaitForReportedPropertyUpdatesInternalAsync(expected, token),
                 "Received expected twin updates for module '{Module}'",
                 this.Id);
         }
+
+        protected Task WaitForReportedPropertyUpdatesInternalAsync(object expected, CancellationToken token)
+        {
+            return Retry.Do(
+                async () =>
+                {
+                    Twin twin = await this.iotHub.GetTwinAsync(this.deviceId, this.Id, token);
+                    return twin.Properties.Reported;
+                },
+                reported => JsonEquals((expected, "properties.reported"), (reported, string.Empty)),
+                null,
+                TimeSpan.FromSeconds(5),
+                token);
+        }
+
+        // reference.rootPath and comparand.rootPath are path strings like those returned from
+        // Newtonsoft.Json.Linq.JToken.Path, and compatible with the path argument to
+        // Newtonsoft.Json.Linq.JToken.SelectToken(path)
+        static bool JsonEquals(
+            (object obj, string rootPath) reference,
+            (object obj, string rootPath) comparand)
+        {
+            Dictionary<string, JValue> ProcessJson(object obj, string rootPath)
+            {
+                // return all json values under root path, with their relative
+                // paths as keys
+                int length = rootPath.Length + (rootPath == string.Empty ? 0 : 1);
+                return JObject
+                    .FromObject(obj)
+                    .SelectToken(rootPath)
+                    .Cast<JContainer>()
+                    .DescendantsAndSelf()
+                    .OfType<JValue>()
+                    .ToDictionary(v => v.Path.Substring(length));
+            }
+
+            var referenceValues = ProcessJson(reference.obj, reference.rootPath);
+            var comparandValues = ProcessJson(comparand.obj, comparand.rootPath);
```

Comment:
Recommend using explicit types instead of var

Reply:
Oh yeah meant to go back and do that. I'll make a pass.

Reply:
Done.